### ISSUE TITLE
amp-bind: Support intermediary vars in AMP.setState()

### ISSUE THREAD

**Issue by choumx** [ampproject/amphtml#8397]
### Benefits

- Reduces document markup due to code duplication.
- Improves performance by caching expression fragments.

### Possible solution

One approach is to allow chained actions. E.g. instead of...
```
on="tap:AMP.setState({foo: 1 + 2, bar: (1 + 2) + 3})"
```
...do this:
```
on="tap:AMP.setState({onePlusTwo: 1 + 2}), 
        AMP.setState({foo: onePlusTwo, bar: onePlusTwo + 3})"
```
This doesn't work today because the second action doesn't wait for the first to complete.

Only really painful in more complicated use cases though, e.g. `examples/bind/turing.amp.html`.

**Comment by kmh287** [comment]
Could setState take a variable number of object literals and then call bind's setState on them in order?

**Comment by kmh287** [comment]
Also, couldn't this be used to get around our expression complexity limit?

**Comment by choumx** [comment]
> Could setState take a variable number of object literals and then call bind's setState on them in order?

Yes, though it's a slightly more complex change since we'd have to change the action parser.

> Also, couldn't this be used to get around our expression complexity limit?

Expressions in `setState` should go through the same code path.

**Comment by kmh287** [comment]
I'm thinking though that if we allow multi-step calculations, then someone could just break up a larger calculation into multiple steps to circumvent the complexity limit.

For instance if my expression is A + B where (A + B) has an AST size of 51 (one over the current limit) then I could do.

`AMP.setState({x: A, y: B}), AMP.setState({z: x+y})`

**Comment by choumx** [comment]
Good point, we'd have to have a more sophisticated constraint to avoid that.

**Comment by kmh287** [comment]
We could put a limit on the number of sequential setStates, but even then each one we allow multiplies our effective complexity limit.

**Comment by choumx** [comment]
Another option is to add a new custom element:
```
<amp-variable id="onePlusOne" expression="1 + 1">
```
Slightly less powerful and more explicit than chaining `AMP.setState` actions.

**Comment by kmh287** [comment]
Once #8678 is in, won't users be able to chain `setState`?

**Comment by choumx** [comment]
Fixed with #12285.
